src/components/Requirements/IncompleteSelfCheck.vue

```diff
@@ -0,0 +1,260 @@
+  },
+  props: {},
+  data(): Data {
+    return {
+      showDropdown: false,
+      isCourseModalOpen: false,
+    };
+  },
+  computed: {
+    selfCheckCourses(): Record<string, FirestoreSemesterCourse> {
+      // TODO - limit to courses that don't have requirements they are fulfilling
+      const courses: Record<string, FirestoreSemesterCourse> = {};
+      store.state.semesters.forEach(semester => {
+        semester.courses.forEach(course => {
+          courses[course.code] = course;
+        });
+      });
+      return courses;
+    },
+  },
+  directives: {
+    'click-outside': clickOutside,
+  },
+  methods: {
+    closeMenuIfOpen() {
+      this.showDropdown = false;
+    },
+    addExistingCourse(option: string) {
+      this.showDropdown = false;
+      this.$emit('addCourse', this.selfCheckCourses[option]);
+    },
+    addNewCourse(course: CornellCourseRosterCourse, season: FirestoreSemesterType, year: number) {
+      this.showDropdown = false;
+      const newCourse = cornellCourseRosterCourseToFirebaseSemesterCourse(course);
+      this.addCourseToSemester(season, year, newCourse);
+      this.$emit('addCourse', newCourse);
+    },
+    openCourseModal() {
+      this.isCourseModalOpen = true;
+    },
+    closeCourseModal() {
+      this.isCourseModalOpen = false;
+    },
+    checkCourseDuplicate(key: string) {
+      // if (this.courses) {
+      //   // @ts-ignore
+      //   this.$refs.modal.courseIsAddable = true;
+      //   this.courses.forEach(course => {
+      //     if (course.code === key) {
+      //       // @ts-ignore
+      //       this.$refs.modal.courseIsAddable = false;
+      //       this.$emit('open-caution-modal');
+      //     }
+      //   });
+      // }
+    },
+
+    addCourseToSemester(
+      season: FirestoreSemesterType,
+      year: number,
+      newCourse: FirestoreSemesterCourse
+    ) {
+      editSemesters(oldSemesters => {
+        let semesterFound = false;
+        const newSemestersWithCourse = oldSemesters.map(sem => {
+          if (sem.type === season && sem.year === year) {
+            semesterFound = true;
+            return { ...sem, courses: [...sem.courses, newCourse] };
+          }
+          return sem;
+        });
+        if (semesterFound) return newSemestersWithCourse;
+        return [...oldSemesters, this.createSemester([newCourse], season, year)].sort(this.compare);
+      });
+    },
+    createSemester(
+      courses: readonly FirestoreSemesterCourse[],
+      type: FirestoreSemesterType,
+      year: number
+    ) {
+      return { courses, type, year };
+    },
+    compare(a: FirestoreSemester, b: FirestoreSemester): number {
+      if (a.type === b.type && a.year === b.year) {
+        return 0;
+      }
+      if (a.year > b.year) {
+        return -1;
+      }
+      if (a.year < b.year) {
+        return 1;
+      }
+      if (SeasonsEnum[a.type] < SeasonsEnum[b.type]) {
+        return 1;
+      }
+      return -1;
+    },
+  },
```

Comment:
Maybe we can put the duplicated functions in a utils file. I've also seen `SeasonsEnum` several times in the codebase.